Align thread priority with Linux defaults

behlendorf opened the pull request:
Creating threads with maxclsyspri under Linux results in threads
with lower than the default kernel thread priority.  This means
that on an active system they can be starved for CPU cycles which
negatively impacts performance.  To prevent this from happening
defclsyspri should be used so they are created with the default
system priority.

Signed-off-by: Brian Behlendorf <behlendorf1@llnl.gov>
Issue #3607

ryao commented:
This is probably because higher numbers are higher priority on Solaris while lower numbers are lower priority on Linux. We probably should change maxclsyspri to `NICE_TO_PRIO(MIN_NICE)`.

http://lxr.free-electrons.com/source/include/linux/sched/prio.h

FransUrbo commented:
@ryao You mean the other way around, right - "lower numbers are higher priority on Linux"? :)

behlendorf commented:
@ryao that's not a bad idea although it's probably not quite enough.  There are places in the code where the priority gets incremented or decremented and those would need to be inverted as well.  For the moment, I've just added a `defclsyspri` to the spl code and used that as the default.

I'm sorely tempted to just move everything to the default until we have data for specific taskq's which suggest better values for Linux.  There's no great reason to believe the illumos values are good choices for Linux considering how different the schedulers are.

FransUrbo commented:
I think the default for filesystem "stuff" is somewhere between <code>-5</code> and <code>-10</code>, <code>0</code> being the normal default for ordinary processes etc.

kernelOfTruth commented:
FYI:

the following is with 4.1 kernel:

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep xfs
     -  TS   264  39 -20 [xfsalloc]
     -  TS   265  39 -20 [xfs_mru_cache]
```

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep jfs
     -  TS   254  19   0 [jfsIO]
     -  TS   255  19   0 [jfsCommit]
     -  TS   256  19   0 [jfsCommit]
     -  TS   257  19   0 [jfsCommit]
     -  TS   258  19   0 [jfsCommit]
     -  TS   259  19   0 [jfsCommit]
     -  TS   260  19   0 [jfsCommit]
     -  TS   261  19   0 [jfsCommit]
     -  TS   262  19   0 [jfsCommit]
     -  TS   263  19   0 [jfsSync]
```

```ps -eLo rtprio,cls,pid,pri,nice,cmd | grep crypt
     -  TS    56  39 -20 [crypto]
     -  TS   287  39 -20 [pencrypt]
     -  TS   288  39 -20 [pdecrypt]
     -  TS  1780  39 -20 [kcryptd_io]
     -  TS  1781  39 -20 [kcryptd]
     -  TS  1782  19   0 [dmcrypt_write]
     -  TS 13527  39 -20 [kcryptd_io]
     -  TS 13530  39 -20 [kcryptd]
     -  TS 13531  19   0 [dmcrypt_write]
     -  TS 19821  39 -20 [kcryptd_io]
     -  TS 19823  39 -20 [kcryptd]
     -  TS 19824  19   0 [dmcrypt_write]
     -  TS 25845  39 -20 [kcryptd_io]
     -  TS 25847  39 -20 [kcryptd]
     -  TS 25848  19   0 [dmcrypt_write]
     -  TS 26064  39 -20 [kcryptd_io]
     -  TS 26065  39 -20 [kcryptd]
     -  TS 26067  19   0 [dmcrypt_write]
     -  TS 26714  39 -20 [kcryptd_io]
     -  TS 26715  39 -20 [kcryptd]
     -  TS 26717  19   0 [dmcrypt_write]
     -  TS 26729  39 -20 [kcryptd_io]
     -  TS 26730  39 -20 [kcryptd]
     -  TS 26731  19   0 [dmcrypt_write]
```

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep kwork
     -  TS     5  39 -20 [kworker/0:0H]
     -  TS    17  39 -20 [kworker/1:0H]
     -  TS    22  39 -20 [kworker/2:0H]
     -  TS    27  39 -20 [kworker/3:0H]
     -  TS    32  39 -20 [kworker/4:0H]
     -  TS    38  39 -20 [kworker/5:0H]
     -  TS    43  39 -20 [kworker/6:0H]
     -  TS    48  39 -20 [kworker/7:0H]
     -  TS   844  39 -20 [kworker/2:1H]
     -  TS  1837  39 -20 [kworker/6:1H]
     -  TS  1858  39 -20 [kworker/5:1H]
     -  TS  1861  39 -20 [kworker/0:1H]
     -  TS  1863  39 -20 [kworker/3:1H]
     -  TS  1865  39 -20 [kworker/4:1H]
     -  TS  1894  39 -20 [kworker/1:1H]
     -  TS  1959  39 -20 [kworker/7:1H]
     -  TS  9305  39 -20 [kworker/u17:0]
     -  TS 12974  19   0 [kworker/2:3]
     -  TS 22203  19   0 [kworker/6:2]
     -  TS 22503  19   0 [kworker/4:1]
     -  TS 23031  19   0 [kworker/1:2]
     -  TS 23184  19   0 [kworker/u16:6]
     -  TS 23424  39 -20 [kworker/u17:1]
     -  TS 23543  19   0 [kworker/6:0]
     -  TS 23665  19   0 [kworker/u16:3]
     -  TS 23811  19   0 [kworker/5:2]
     -  TS 24161  19   0 [kworker/7:1]
     -  TS 24317  19   0 [kworker/1:1]
     -  TS 24474  19   0 [kworker/u16:0]
     -  TS 24480  19   0 [kworker/u16:1]
     -  TS 24481  19   0 [kworker/u16:2]
     -  TS 24482  19   0 [kworker/u16:4]
     -  TS 24483  19   0 [kworker/u16:5]
     -  TS 24484  19   0 [kworker/u16:7]
     -  TS 24558  19   0 [kworker/1:0]
     -  TS 24657  19   0 [kworker/5:1]
     -  TS 24677  19   0 [kworker/7:0]
     -  TS 24730  19   0 [kworker/3:0]
     -  TS 24747  19   0 [kworker/4:0]
     -  TS 25469  19   0 [kworker/3:1]
     -  TS 25560  19   0 [kworker/0:1]
     -  TS 27151  19   0 [kworker/7:3]
     -  TS 28346  19   0 [kworker/0:4]
     -  TS 28349  19   0 [kworker/2:1]
     -  TS 29018  19   0 [kworker/5:0]
     -  TS 31092  19   0 [kworker/4:2]
     -  TS 31797  19   0 [kworker/3:2]
```

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep btrfs
     -  TS  1810  39 -20 [btrfs-worker]
     -  TS  1812  39 -20 [btrfs-worker-hi]
     -  TS  1813  39 -20 [btrfs-delalloc]
     -  TS  1814  39 -20 [btrfs-flush_del]
     -  TS  1815  39 -20 [btrfs-cache]
     -  TS  1816  39 -20 [btrfs-submit]
     -  TS  1817  39 -20 [btrfs-fixup]
     -  TS  1818  39 -20 [btrfs-endio]
     -  TS  1819  39 -20 [btrfs-endio-met]
     -  TS  1820  39 -20 [btrfs-endio-met]
     -  TS  1821  39 -20 [btrfs-endio-rai]
     -  TS  1822  39 -20 [btrfs-endio-rep]
     -  TS  1823  39 -20 [btrfs-rmw]
     -  TS  1824  39 -20 [btrfs-endio-wri]
     -  TS  1825  39 -20 [btrfs-freespace]
     -  TS  1826  39 -20 [btrfs-delayed-m]
     -  TS  1827  39 -20 [btrfs-readahead]
     -  TS  1828  39 -20 [btrfs-qgroup-re]
     -  TS  1829  39 -20 [btrfs-extent-re]
     -  TS  1830  19   0 [btrfs-cleaner]
     -  TS  1831  19   0 [btrfs-transacti]
     -  TS 14843  39 -20 [btrfs-worker]
     -  TS 14845  39 -20 [btrfs-worker-hi]
     -  TS 14846  39 -20 [btrfs-delalloc]
     -  TS 14847  39 -20 [btrfs-flush_del]
     -  TS 14848  39 -20 [btrfs-cache]
     -  TS 14849  39 -20 [btrfs-submit]
     -  TS 14850  39 -20 [btrfs-fixup]
     -  TS 14851  39 -20 [btrfs-endio]
     -  TS 14852  39 -20 [btrfs-endio-met]
     -  TS 14853  39 -20 [btrfs-endio-met]
     -  TS 14854  39 -20 [btrfs-endio-rai]
     -  TS 14855  39 -20 [btrfs-endio-rep]
     -  TS 14856  39 -20 [btrfs-rmw]
     -  TS 14857  39 -20 [btrfs-endio-wri]
     -  TS 14858  39 -20 [btrfs-freespace]
     -  TS 14859  39 -20 [btrfs-delayed-m]
     -  TS 14860  39 -20 [btrfs-readahead]
     -  TS 14861  39 -20 [btrfs-qgroup-re]
     -  TS 14862  39 -20 [btrfs-extent-re]
     -  TS 14867  19   0 [btrfs-cleaner]
     -  TS 14868  19   0 [btrfs-transacti]
     -  TS 26641  39 -20 [btrfs-worker]
     -  TS 26642  39 -20 [btrfs-worker-hi]
     -  TS 26643  39 -20 [btrfs-delalloc]
     -  TS 26644  39 -20 [btrfs-flush_del]
     -  TS 26645  39 -20 [btrfs-cache]
     -  TS 26646  39 -20 [btrfs-submit]
     -  TS 26647  39 -20 [btrfs-fixup]
     -  TS 26648  39 -20 [btrfs-endio]
     -  TS 26649  39 -20 [btrfs-endio-met]
     -  TS 26650  39 -20 [btrfs-endio-met]
     -  TS 26651  39 -20 [btrfs-endio-rai]
     -  TS 26652  39 -20 [btrfs-endio-rep]
     -  TS 26653  39 -20 [btrfs-rmw]
     -  TS 26654  39 -20 [btrfs-endio-wri]
     -  TS 26655  39 -20 [btrfs-freespace]
     -  TS 26656  39 -20 [btrfs-delayed-m]
     -  TS 26657  39 -20 [btrfs-readahead]
     -  TS 26658  39 -20 [btrfs-qgroup-re]
     -  TS 26659  39 -20 [btrfs-extent-re]
     -  TS 26660  19   0 [btrfs-cleaner]
     -  TS 26661  19   0 [btrfs-transacti]
```

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep zil
     -  TS 27584  39 -20 [zil_clean]
     -  TS 27589  39 -20 [zil_clean]
     -  TS 27592  39 -20 [zil_clean]
     -  TS 27595  39 -20 [zil_clean]
     -  TS 27598  39 -20 [zil_clean]
     -  TS 27601  39 -20 [zil_clean]
     -  TS 27605  39 -20 [zil_clean]
     -  TS 27611  39 -20 [zil_clean]
     -  TS 27615  39 -20 [zil_clean]
     -  TS 27618  39 -20 [zil_clean]
     -  TS 27621  39 -20 [zil_clean]
     -  TS 27624  39 -20 [zil_clean]
     -  TS 27627  39 -20 [zil_clean]
     -  TS 27630  39 -20 [zil_clean]
     -  TS 27633  39 -20 [zil_clean]
     -  TS 27636  39 -20 [zil_clean]
     -  TS 27639  39 -20 [zil_clean]
     -  TS 27642  39 -20 [zil_clean]
     -  TS 27646  39 -20 [zil_clean]
     -  TS 27649  39 -20 [zil_clean]
     -  TS 27652  39 -20 [zil_clean]
     -  TS 27655  39 -20 [zil_clean]
     -  TS 27658  39 -20 [zil_clean]
     -  TS 27661  39 -20 [zil_clean]
     -  TS 27664  39 -20 [zil_clean]
     -  TS 27667  39 -20 [zil_clean]
...
```


```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep spl
     -  TS 24864  19   0 grep --color=auto spl
     -  TS 26976   0  19 [spl_kmem_cache]
     -  TS 26977  39 -20 [spl_system_task]
     -  TS 26978  39 -20 [spl_dynamic_tas]
```

```
ps -eLo rtprio,cls,pid,pri,nice,cmd | grep z_
     -  TS 24880   1  18 [z_wr_iss]
     -  TS 24881   1  18 [z_wr_iss]
     -  TS 24882   1  18 [z_wr_iss]
     -  TS 24884   1  18 [z_wr_iss]
     -  TS 26984   0  19 [z_unmount]
     -  TS 27191   0  19 [z_null_iss]
     -  TS 27192   0  19 [z_null_int]
     -  TS 27193   0  19 [z_rd_iss]
     -  TS 27194   0  19 [z_rd_int_0]
     -  TS 27195   0  19 [z_rd_int_1]
     -  TS 27196   0  19 [z_rd_int_2]
     -  TS 27197   0  19 [z_rd_int_3]
     -  TS 27198   0  19 [z_rd_int_4]
     -  TS 27199   0  19 [z_rd_int_5]
     -  TS 27200   0  19 [z_rd_int_6]
     -  TS 27201   0  19 [z_rd_int_7]
     -  TS 27202   1  18 [z_wr_iss]
     -  TS 27203   0  19 [z_wr_iss_h]
     -  TS 27204   0  19 [z_wr_int_0]
     -  TS 27205   0  19 [z_wr_int_1]
     -  TS 27206   0  19 [z_wr_int_2]
     -  TS 27207   0  19 [z_wr_int_3]
     -  TS 27208   0  19 [z_wr_int_4]
     -  TS 27209   0  19 [z_wr_int_5]
     -  TS 27210   0  19 [z_wr_int_6]
     -  TS 27211   0  19 [z_wr_int_7]
     -  TS 27212   0  19 [z_wr_int_h]
     -  TS 27213   0  19 [z_fr_iss_0]
     -  TS 27214   0  19 [z_fr_iss_1]
     -  TS 27215   0  19 [z_fr_iss_2]
     -  TS 27216   0  19 [z_fr_iss_3]
     -  TS 27217   0  19 [z_fr_iss_4]
     -  TS 27218   0  19 [z_fr_iss_5]
     -  TS 27219   0  19 [z_fr_iss_6]
     -  TS 27220   0  19 [z_fr_iss_7]
     -  TS 27221   0  19 [z_fr_int]
     -  TS 27222   0  19 [z_cl_iss]
     -  TS 27223   0  19 [z_cl_int]
     -  TS 27224   0  19 [z_ioctl_iss]
     -  TS 27225   0  19 [z_ioctl_int]
     -  TS 27230  39 -20 [z_iput]
```

behlendorf commented:
Requires zfsonlinux/zfs#466

@kernelOfTruth thanks for pointing out what other common Linux filesystems use for process priority.  I've refreshed this patch to use those values as a base line.  While changing the threads to the default priority was good enough keep user space processes from negatively impacting performance.  It occurs to me that it won't be good enough to prevent another active native Linux filesystem from potentially starving ZFS performance.  We want them both to be running with the same priority if possible.  I've also integrated @ryao's idea of inverting the min/max values.

I'd be very interested to see how these patches impact performance on busy systems with real workloads.

kernelOfTruth commented:
@behlendorf glad to be of help

https://github.com/zfsonlinux/spl/pull/466 =)

behlendorf commented:
Whoops!  I corrected the comment to point to the correct issue.

nedbass commented:
LGTM. The choices seem reasonable and @FransUrbo makes good points about future-proofing this a bit.

behlendorf commented:
Refreshed with the follow changes.

* Updated the commit comment
* Extend the comment above `pri++` to explain why it differs from illumos
* Added support to set the priority in the user space taskq_create()  and thread_create().

@FransUrbo regarding the specific priority of the tasks they'll all listed in the commit message.  But the general rule I went by was any task that might perform I/O should have the maximum priority.  Other helper threads which are less critical should run with the default.  It's important that they run but it's less critical.  Incidentally this is far better than the current situation where everything is running with a very low priority.

FransUrbo commented:
I still think that <code>defclsyspri</code> should be a negative value. Maybe only something like <code>-2</code>. But on the other hand, I'm not entirely sure what those threads do. Are they really as (un)important as say <code>httpd</code>, <code>syslogd</code> and what not?

Since they're (in one way or the other) associated with a filesystem, I would have expected them to be (slightly) more important… Document why they're not?

Also, the additional comment kind'a missed my point (slightly) I was trying to make…

behlendorf commented:
@FransUrbo the following threads are all `defclsyspri` and are largely responsible for managing the cache asynchronously.  None of them are time critical like an I/O because normally no user task will be blocked waiting on them.  They just need to run fairly promptly to get their work done.  We can definitely redefine their default values going forward if we find it important but I think starting with the default priority is a good place to start.

__dbu_evict__ - Asynchronously evicting user buffers assoicated with a dbuf.
__arc_user_evicts__ - Asynchronously evicting user buffers assoicated with arc buffers
__arc_prune__ - Proactively dropping cached inodes/dnodes as may be required.
__arc_reclaim__ - Proactively managing arc_c/arc_p targets.
__l2arc_feed__ - Lazily filling the L2ARC cache from the primary ARC.
__z_unmount__ - Automounts .zfs snapshots
__z_iput__ - Asynchronously drops the last inode reference in a deadlock-free way.
```

> Also, the additional comment kind'a missed my point (slightly) I was trying to make…

I must have misunderstood your point.  Care to suggest something else?

FransUrbo commented:
> @FransUrbo the following threads are all defclsyspri and are largely responsible for managing the cache asynchronously.
> 
Ok, could we document this as well with a online comment?

Something like:

	-	dp->dp_iput_taskq = taskq_create("z_iput", max_ncpus, minclsyspri,
	+	/* Asynchronously drops the last inode reference in a deadlock-free way. */
	+	dp->dp_iput_taskq = taskq_create("z_iput", max_ncpus, defclsyspri,

I know that this is somewhat out of scope for the PR, but still… While the code is modified, why not just do it the whole way?
> > Also, the additional comment kind'a missed my point (slightly) I was trying to make…
> 
> I must have misunderstood your point.  Care to suggest something else?
That it would be easy to find this.

For this, something like:

	/*
	 * LINUX: Under Linux this means incrementing the priority value.
	 *              On platforms like Illumos it should be decremented.
	 */

That way, whoever wants to try to merge the Illumos and Linux code (etc) - like I was about to start a couple of months ago, can easily find what is very Linux specific and shouldn't be touched and merged with Illumos. I couldn't even begin, because I didn't know what's Linux unique, what haven't been ported yet and what's "somewhere in between" (?).

	rgrep LINUX: .

I bet no-one, not even you guys that fiddle with the core code every day, can say exactly what's Linux unique without missing one (or several). This way, it's absolutly clear to anyone and everyone that this is Linux-unique and not just something we haven't ported yet.

As it is now, only by knowing every line of code, going through it AND knowing the history of the code could one find "this stuff". With a comment like that (if you think a "#ifdef LINUX" is to much), it's easy to find and anyone looking through the code will recognize it, even in years to come when Illumos "fiddles" with this...

behlendorf commented:
I think if we're going to start commenting what every taskq is for we should do it in a separate patch and include more than a one line comment.  Let's leave that for another time.  Keeping it in a separate commit would also make it easier to push upstream which is when we'd ideally like it to be.

As for the additional Linux comment we'll have to disagree about the style.  Having something you can grep for I don't think is the most important part.  Having an explanation for why something is different is what you really want to know.  Ideally we want to hide any differences behind a wrapper function to keep the code as readable as possible.  I think the updated comment accomplished that.

FransUrbo commented:
You've always been over thinking things :). "**A journey of a thousand miles starts with a single step**".

Yes, finding all of them and add comments in one (or several commits) and then push it upstream IS preferable. But starting with this one, here and now is a very good start. NO ONE is going to have neither the time, interest or will to do the "full monty", and you know it… Hence, "it's never going to happen". Think small for once.

behlendorf commented:
Merged as:

1229323 Align thread priority with Linux defaults
